In this experiment I am going to check for separation between various groups based on the EMT gene signature from Byers et al., 2013. This signature includes both mesenchymal and epithelial cell markers, so combined should be able to distinguish between the cell types (of note: this signature was developed in tumor data)
  
```{r defineFunctions, include=FALSE}
foldChange <- function(eset, groups, genes){
  apply(exprs(eset)[genes, groups==0], 1, mean) - 
    apply(exprs(eset)[genes, groups==1], 1, mean)
}

medianFilter <- function(eset){
  md <- median(exprs(eset))
  passFilter <- logical(featureNumber(eset))
  for(i in 1:featureNumber(eset)){
    passFilter[i] <- sum(exprs(eset)[i, ] > md) > 0
  }
  
  eset <- eset[passFilter, ]
  return(eset)
  
}

saveGeneList <- function(analysis, filename=paste(Sys.Date(), "Temp", analysis$name, ".txt", sep="")){
  
  write.table(analysis$geneSymbols, file=filename, quote=FALSE,row.names=FALSE, col.names=FALSE)

}

saveRankedList <- function(fit, varColInd, filename = "tempRankedList"){
  filename = paste(paste(Sys.Date(), colnames(fit$t)[varColInd], filename, sep="_"), "rnk", sep=".")
  print(filename)
  temp <- data.frame(geneSymbols = fit$genes$Symbol, ts = fit$t[, varColInd])
  write.table(temp, file=filename, row.names=FALSE, quote=FALSE, col.names=TRUE, sep="\t", eol="\n")
  
  
}

saveGeneSet <- function(fit, inds, filename = "tempGeneSet", geneSetName="GeneSet", description=""){
  filename = paste(paste(Sys.Date(), filename, sep="_"), "gmx", sep=".")
  print(filename)
  temp <- data.frame(geneSymbols=c(geneSetName, description, fit$genes$Symbol[inds]))
  write.table(temp, file=filename, row.names=FALSE, quote=FALSE, col.names=FALSE)
  
}


```

### Setup
```{r setup, eval=TRUE, echo=TRUE, include=FALSE, results='hide'}

setwd("/protected/projects/pulmarray/Allegro/COPD_Cancer/experiments/Expmt_0041_Cancer_Comparisons/")
source("/protected/projects/pulmarray/Allegro/COPD_Cancer/scripts/AllegroSetup.R")
#source("../2015-05-03_CBM_ATS_2015/plotGSEA.R")
# fix the one patient with wonky data
# eventually this should just be saved in the RDS file
holdEset <- eset
holdEset$FEV1Pc[holdEset$FEV1Pc==89.2] <- 0.892

emt_genes <- read.csv("/protected/projects/pulmarray/Allegro/COPD_Cancer/geneSubsets/emt_gene_symbols_byers_2013.csv", header=FALSE)
temp <- as.factor(unique(as.character(emt_genes$V1)))
emt_genes <- data.frame("EMT_Genes" = temp)

GOLD_colors = c("0" = "gray100", "1" = "gray75", "2" = "gray50", "3" = "gray25", "4" = "black")
greyblack <- colorRampPalette(c("black", "white", "grey"))(256)


# load ATF4 BEASB2 gene perturbation data for analysis
atf4_data <- readRDS("/protected/projects/pulmarray/Allegro/ATF4/ATF4data_2015MAR17_ExpressionSet.rds")

lam_sign <- read.delim("/protected/projects/pulmarray/Allegro/COPD_Cancer/geneSubsets/katie98genesEntrez.txt")
lam_sign[, 3] <- paste(lam_sign[, 2], "_at", sep="")

lamGMT <- read.csv("/protected/projects/pulmarray/Allegro/COPD_Cancer/geneSubsets/100525_lam_98overlapgenes_covars.gmt", sep="\t", as.is=TRUE, header=FALSE)
lamUP <- as.character(lamGMT[1, 3:56])
lamDN <- as.character(lamGMT[2, 3:46])

joshLM127dn <- read.csv("/protected/projects/pulmarray/HoggCOPD/A1ATGrifols/experiments/Expmt_0001_LM_Parenchyma/2015-06-08_Joshdn127.gmx", skip=1, header=TRUE)

joshLM127up <- read.csv("/protected/projects/pulmarray/HoggCOPD/A1ATGrifols/experiments/Expmt_0001_LM_Parenchyma/2015-06-08_Joshup127.gmx", skip=1, header=TRUE)

# from parenchymal REAL data
cleLMup <- read.csv("/protected/projects/pulmarray/HoggCOPD/A1ATGrifols/experiments/Expmt_0003_LM_Parenchyma_CLE/2015-06-24_parenchyma_cle_geneset_cluster2_fdr_0.05.gmx")

cleLMdn <- read.csv("/protected/projects/pulmarray/HoggCOPD/A1ATGrifols/experiments/Expmt_0003_LM_Parenchyma_CLE/2015-06-24_parenchyma_cle_geneset_cluster1_fdr_0.05.gmx")

natMed <- read.csv("/protected/projects/pulmarray/Allegro/COPD_Cancer/geneSubsets/Nature_Medicine_gmx_files/top_100_genes_symbols_quality_adjusted.gmx", sep="\t", as.is=TRUE)
natMedUp <- natMed[2:65, 1]
natMedDn <- natMed[2:37, 2]

tgfB <- read.csv("/protected/projects/pulmarray/Allegro/COPD_Cancer/geneSubsets/msigdb_C2_CGP_genesets/BIOCARTA_TGFB_PATHWAY.gmx", skip=1)
tgfB <- as.character(tgfB$X..TGF.beta.signaling.pathway)



#atf4_targets <- read.delim("ATF4_Targets_Enrichr_TRANSFACJASPAR_270Genes.txt", sep=",", as.is=TRUE, header=FALSE)
#atf4_targets <- as.character(atf4_targets)

#a4LEup <- read.delim("ATF4_targets_cluster1_270Genes_LeadingEdge_Up.csv", header=FALSE)
#a4LEup <- as.character(a4LEup$V1)

#a4LEdn <- read.delim("ATF4_targets_cluster2_270Genes_LeadingEdge_Down.csv", header=FALSE)
#a4LEdn <- as.character(a4LEdn$V1)
library(znorm)
library(diptest)
library(GSVA)
library(annotate)
library(org.Hs.eg.db)
library(corrplot)
library(ggplot2)
library(beeswarm)


```

### Data Cleaning
```{r cleanup}

# Remaining question - should we keep or remove the COPD GOLD Status 1 patients? i.e. mild disease
# What about patients using or their charts indicating they have used corticosteroids?
esetClean <- holdEset
esetClean <- cleanNAForAnalysis(esetClean, "COPD2_R7")
esetClean <- cleanNAForAnalysis(esetClean, "RATIOc")
esetClean <- cleanNAForAnalysis(esetClean, "PYc")
esetClean <- cleanNAForAnalysis(esetClean, "RIN")
esetClean <- removeFactorLevel(esetClean, "COPD2_R7", "DK")
esetClean <- removeFactorLevel(esetClean, "FinalCaDXc", "DK")
esetClean <- removeFactorLevel(esetClean, "COPD2_R7", "DK")
esetClean <- removeFactorLevel(esetClean, "GENDERc", "DK")

esetClean <- calcIndicator(esetClean, "FinalCaDXc", "COPD2_R7")
esetClean$smkindic <- as.numeric(as.character(esetClean$indicator))
esetClean$smkindic[esetClean$SMKc==2] <- esetClean$smkindic[esetClean$SMKc==2] + 4
esetClean$smkindic <- as.factor(esetClean$smkindic)

# Define the GOLD Stage appropriate for all patients
esetClean$GOLD <- factor(x=rep(0, sampleNumber(esetClean)), levels=c("0","1","2","3","4"))
# GOLD Stage 1 Ratio < 0.7, FEV1 >= 8
esetClean$GOLD[esetClean$RATIOc < 0.7 & esetClean$FEV1Pc >= 0.8] <- 1
# GOLD Stage 2 Ratio < 0.7, 50 <= FEV1 < 80
esetClean$GOLD[esetClean$RATIOc < 0.7 & esetClean$FEV1Pc < 0.8] <- 2
# GOLD Stage 3 Ratio < 0.7, 30 <= FEV1 <= 50
esetClean$GOLD[esetClean$RATIOc < 0.7 & esetClean$FEV1Pc < 0.5] <- 3
# GOLD Stage 4 Ratio < 0.7, FEV1 < 30%
esetClean$GOLD[esetClean$RATIOc < 0.7 & esetClean$FEV1Pc < 0.3] <- 4

esetCleanFormers <- removeFactorLevel(esetClean, "SMKc", "1")
esetCleanCurrents <- removeFactorLevel(esetClean, "SMKc", "2")
# Generate a super clean dataset devoid of all patients with other pulmonary diseases
esetSuperClean <- esetClean
esetSuperClean <- removeFactorLevel(esetSuperClean, "AllegroPNAc", "1")
esetSuperClean <- removeFactorLevel(esetSuperClean, "AllegroTBc", "1")
esetSuperClean <- removeFactorLevel(esetSuperClean, "AllegroSarcoidc", "1")
esetSuperClean <- removeFactorLevel(esetSuperClean, "AllegroIPFc", "1")
esetSuperClean <- removeFactorLevel(esetSuperClean, "AllegroIPFc", "DK")
esetSuperClean <- removeFactorLevel(esetSuperClean, "AllegroAtelectasisc", "1")
esetSuperClean <- removeFactorLevel(esetSuperClean, "AllegroAtelectasisc", "DK")
esetSuperClean <- removeFactorLevel(esetSuperClean, "AllegroAsthmac", "DK")
esetSuperClean <- removeFactorLevel(esetSuperClean, "AllegroAsthmac", "1")

esetSuperCleanICS <- removeFactorLevel(esetSuperClean, "AllegroCorticosteroidsc", "1")
esetSuperCleanICS <- removeFactorLevel(esetSuperCleanICS, "AllegroCorticosteroidsc", "DK")

# clean up ATF4 data
varLabels(atf4_data) <- "SampleID"
atf4_data$SampleID[grep("ATF4", atf4_data$SampleID)] <- "ATF4"
atf4_data$SampleID[grep("Ctrl", atf4_data$SampleID)] <- "Control"
atf4_data$SampleID <- as.factor(atf4_data$SampleID)

atf4_colors <- c("ATF4" = "red", "Control" = "black")


```

### ATF4 Models
```{r ATF4models}
# generate ATF4 model(s)
atf4Model <- lmFitWrapper(atf4_data, c("SampleID"),
                          name="ATF4 Perturbation Data", adjust.method="none",
                          p.value=0.05, varOfInterest=1)

clabels <- cbind(atf4_colors[as.character(atf4_data$SampleID)])
heatmap3(exprs(atf4_data)[atf4Model$inds, ], col = bluered,
         hclustfun=function(d) hclust(d, method="ward.D"),
         col.clustering = atf4_data$SampleID, ColSideColors = clabels,
         main = "ATF4 Genes",
         labCol="")

hmapClusts <- heatmap3(exprs(atf4_data)[atf4Model$inds,], col = bluered,
                       keep.dendro=TRUE,
                       hclustfun=function(d) hclust(d, method="ward.D"),
                       col.clustering = atf4_data$SampleID, ColSideColors=clabels,
                       main="ATF4 Genes", labCol="")

clusters <- cutree(as.hclust(hmapClusts$Rowv), k=2)
atf4Cluster1_dn <- atf4Model$geneSymbols[clusters==1]
atf4Cluster2_up <- atf4Model$geneSymbols[clusters==2]

atf4dn <- names(clusters)[clusters==1]
atf4up <- names(clusters)[clusters==2]
atf4 <- names(clusters)

rlabels <- cbind(smoking_colors[clusters])
heatmap3(exprs(atf4_data)[atf4Model$inds,], col = bluered,
         keep.dendro=TRUE,
         hclustfun=function(d) hclust(d, method="ward.D"),
         col.clustering = atf4_data$SampleID, ColSideColors=clabels,
         main="ATF4 Genes", labCol="", RowSideColors=rlabels)



```

## Cancer in non-COPD
### Genes changing with cancer in former Smokers
```{r cancerGenesFormer}
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetCleanFormers), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "1")
cccc <- lmFitWrapper(eset, c("FinalCaDXc", "AllegroCorticosteroidsc", "GENDERc", "RATIOc"),1, "none", 0.01)

# generate the gene ranked list for genes changing with cancer among non-COPD-ers
#rankedList <- data.frame(genes=cccc$fit$genes$Symbol, tstat=cccc$fit$t[,2])

# write out the ranked list to file
filename = "nonCOPDCancerFormer"
hmapName = "Genes changing with cancer in non-COPD in Formers"
saveRankedList(cccc$fit, 2, filename=filename)
#filename = "nonCOPDCancerFormerRankedList"
#filename = paste(paste(Sys.Date(), filename, sep="_"), "rnk", sep=".")
#write.table(rankedList, file=filename, row.names=FALSE, quote=FALSE, col.names=TRUE, sep="\t", eol="\n")

clabels <- cbind(cancer_colors[as.character(eset$FinalCaDXc)])
h0 = heatmap3(exprs(eset)[cccc$inds, ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              keep.dendro=TRUE, ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName)

rClusts1 = cutree(as.hclust(h0$Rowv), k=2)
heatmap3(exprs(eset)[cccc$inds, ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1)

clusters <- cutree(as.hclust(h0$Rowv), k=2)
heatmap3(exprs(eset)[cccc$inds[clusters==1], ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1[clusters==1])

fName2 <- paste(filename, "upCancer", sep="_")
saveGeneSet(cccc$fit, cccc$inds[clusters==1], filename=fName2, geneSetName=fName2)

heatmap3(exprs(eset)[cccc$inds[clusters==2], ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1[clusters==2])

fName2 <- paste(filename, "dnCancer", sep="_")
saveGeneSet(cccc$fit, cccc$inds[clusters==2], filename=fName2, geneSetName=fName2)


```

### Genes changing with cancer in current Smokers
```{r cancerGenesCurrent}
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetCleanCurrents), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "1")
cccc <- lmFitWrapper(eset, c("FinalCaDXc", "AllegroCorticosteroidsc", "GENDERc", "RATIOc"),1, "none", 0.01)

# generate the gene ranked list for genes changing with cancer among non-COPD-ers
# write out the ranked list to file
filename = "nonCOPDCancerCurrent"
hmapName = "Genes changing with cancer in non-COPD in Currents"
saveRankedList(cccc$fit, 2, filename=filename)
#filename = "nonCOPDCancerFormerRankedList"
#filename = paste(paste(Sys.Date(), filename, sep="_"), "rnk", sep=".")
#write.table(rankedList, file=filename, row.names=FALSE, quote=FALSE, col.names=TRUE, sep="\t", eol="\n")

clabels <- cbind(cancer_colors[as.character(eset$FinalCaDXc)])
h0 = heatmap3(exprs(eset)[cccc$inds, ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              keep.dendro=TRUE, ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName)

rClusts1 = cutree(as.hclust(h0$Rowv), k=2)
heatmap3(exprs(eset)[cccc$inds, ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1)

clusters <- cutree(as.hclust(h0$Rowv), k=2)
heatmap3(exprs(eset)[cccc$inds[clusters==1], ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1[clusters==1])

fName2 <- paste(filename, "upCancer", sep="_")
saveGeneSet(cccc$fit, cccc$inds[clusters==1], filename=fName2, geneSetName=fName2)

heatmap3(exprs(eset)[cccc$inds[clusters==2], ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1[clusters==2])

fName2 <- paste(filename, "dnCancer", sep="_")
saveGeneSet(cccc$fit, cccc$inds[clusters==2], filename=fName2, geneSetName=fName2)
```

## Cancer in COPD
### Genes changing with cancer in former Smokers
```{r cancerGenesCOPDFormer}
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetCleanFormers), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "0")
cccc <- lmFitWrapper(eset, c("FinalCaDXc", "AllegroCorticosteroidsc", "GENDERc", "RATIOc"),1, "none", 0.005)

# generate the gene ranked list for genes changing with cancer among non-COPD-ers
# write out the ranked list to file
filename = "COPDCancerFormer"
hmapName = "Genes changing with cancer in COPD in Formers"
saveRankedList(cccc$fit, 2, filename=filename)
#filename = "nonCOPDCancerFormerRankedList"
#filename = paste(paste(Sys.Date(), filename, sep="_"), "rnk", sep=".")
#write.table(rankedList, file=filename, row.names=FALSE, quote=FALSE, col.names=TRUE, sep="\t", eol="\n")

clabels <- cbind(cancer_colors[as.character(eset$FinalCaDXc)])
h0 = heatmap3(exprs(eset)[cccc$inds, ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              keep.dendro=TRUE, ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName)

rClusts1 = cutree(as.hclust(h0$Rowv), k=2)
heatmap3(exprs(eset)[cccc$inds, ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1)

clusters <- cutree(as.hclust(h0$Rowv), k=2)
heatmap3(exprs(eset)[cccc$inds[clusters==1], ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1[clusters==1])

fName2 <- paste(filename, "dnCancer", sep="_")
saveGeneSet(cccc$fit, cccc$inds[clusters==1], filename=fName2, geneSetName=fName2)

heatmap3(exprs(eset)[cccc$inds[clusters==2], ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1[clusters==2])

fName2 <- paste(filename, "upCancer", sep="_")
saveGeneSet(cccc$fit, cccc$inds[clusters==2], filename=fName2, geneSetName=fName2)
```

### Genes changing with cancer in current Smokers
```{r cancerGenesCOPDCurrent}
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetCleanCurrents), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "0")
cccc <- lmFitWrapper(eset, c("FinalCaDXc", "AllegroCorticosteroidsc", "GENDERc", "RATIOc"),1, "none", 0.05)

# generate the gene ranked list for genes changing with cancer among non-COPD-ers
# write out the ranked list to file
filename = "COPDCancerCurrent"
hmapName = "Genes changing with cancer in COPD in Currents"
saveRankedList(cccc$fit, 2, filename=filename)
#filename = "nonCOPDCancerFormerRankedList"
#filename = paste(paste(Sys.Date(), filename, sep="_"), "rnk", sep=".")
#write.table(rankedList, file=filename, row.names=FALSE, quote=FALSE, col.names=TRUE, sep="\t", eol="\n")

clabels <- cbind(cancer_colors[as.character(eset$FinalCaDXc)])
h0 = heatmap3(exprs(eset)[cccc$inds, ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              keep.dendro=TRUE, ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName)

rClusts1 = cutree(as.hclust(h0$Rowv), k=2)
heatmap3(exprs(eset)[cccc$inds, ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1)

clusters <- cutree(as.hclust(h0$Rowv), k=2)
heatmap3(exprs(eset)[cccc$inds[clusters==1], ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1[clusters==1])

fName2 <- paste(filename, "upCancer", sep="_")
saveGeneSet(cccc$fit, cccc$inds[clusters==1], filename=fName2, geneSetName=fName2)

heatmap3(exprs(eset)[cccc$inds[clusters==2], ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              ColSideColors=clabels,
              labCol="", col.clustering="semisupervised",
              main=hmapName,
         RowSideColors=rClusts1[clusters==2])

fName2 <- paste(filename, "dnCancer", sep="_")
saveGeneSet(cccc$fit, cccc$inds[clusters==2], filename=fName2, geneSetName=fName2)
```

### Genes changing with cancer in COPD all smokers
```{r copdcancerGenes}
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "0")

tttt <- lmFitWrapper(eset, c("FinalCaDXc", "AllegroCorticosteroidsc", "GENDERc", "SMKc", "RIN"),1, "none", 0.01)

# save out genesets for GSEA comparisons to the non-COPD cancer genes
saveGeneSet(tttt$fit, which(tttt$results[, 2] > 0), filename="copdCancerUpGeneSet", geneSetName="copdCancerUp", description="Genes going up with cancer among patients with COPD")

saveGeneSet(tttt$fit, which(tttt$results[, 2] < 0), filename="copdCancerDnGeneSet", geneSetName="copdCancerDn", description="Genes going down with cancer among patients with COPD")

clabels = cbind(cancer_colors[as.character(eset$FinalCaDXc)], smoking_colors[as.character(eset$SMKc)])
h1 = heatmap3(exprs(eset)[tttt$inds, ], col=bluered,
              hclustfun=function(d) hclust(d, method="ward.D"),
              col.clustering="semisupervised", ColSideColors=clabels,
              labCol="", main="Genes changing with cancer in COPD",
              keep.dendro=TRUE)

clusters = cutree(as.hclust(h1$Rowv), k=2)
ptClusts = cutree(as.hclust(h1$Colv), k=3)
newCols = c("1" = "blue", "2"="red", "3"="green")
clabels = clabels = cbind(cancer_colors[as.character(eset$FinalCaDXc)], newCols[as.character(ptClusts)])
heatmap3(exprs(eset)[tttt$inds, ], col=bluered,
         hclustfun=function(d) hclust(d, method="ward.D"),
         col.clustering="semisupervised", ColSideColors=clabels,
         RowSideColors=clusters, labCol="",
         main="Genes changing with cancer in COPD [clusters]")

# look at these genes changing with cancer in COPD among all patients
esetTemp <- removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK")
clabels <- cbind(copdca_colors[esetTemp$indicator], smoking_colors[esetTemp$SMKc])
heatmap3(exprs(esetTemp)[tttt$inds, ], col=bluered,
         hclustfun=function(d) hclust(d, method="ward.D"),
         col.clustering="semisupervised", ColSideColors=clabels,
         RowSideColors=clusters, labCol="",
         main="Genes changing with cancer in COPD among all patients [clusters]")

heatmap3(exprs(esetTemp)[tttt$inds, ], col=bluered,
         hclustfun=function(d) hclust(d, method="ward.D"),
         col.clustering="unsupervised", ColSideColors=clabels,
         RowSideColors=clusters, labCol="",
         main="Genes changing with cancer in COPD among all patients [clusters]")



```

### Notes on the results from GSEA and enrichr  
1. There is enrichment significantly for each gene set in the opposite direction expected (i.e. those genes going up with cancer in COPD go down with cancer in non-COPD lung cancer; those genes going down with cancer in COPD go up with cancer in non-COPD lung cancer)  
2. Additionally, the genes not included in the core enrichment set for the COPDCancer Up gene set are significantly associated with ZNF217, which has been shown to be important in metastasis free survival in breast and ovarian cancer, but not I believe in lung cancer.  

### Notes on enrichr results for genes up with COPD+Cancer vs COPD  
1. GO Biological Process (16) ERBB signaling pathway (0038127)  
2. GO Biological Process (17) EGFR signaling pathway (0007173)  
3. MSigDB Oncogenic Signature (3) EGFR_UP.V1_DN  
4. MSigDB Oncogenic Signature (5) ERB2_UP.V1_DN  
5. ChEA - YYI (oncogenesis), KDM5B (EMT), STAT3, E2F4, NKX2, E2F1, SOX2, SPI1, FLI1, ZNF217 (EMT)  
6. TRANSFAC/JASPAR - E2F4, E2F1, STAT3, FOXL1, POU1F1, CBEPB, MYB, GATA6, CREM, GATA1, GATA3  
7. TargetScan microRNA MIR493  (regulates E2F1, suupresses tumor growth, invasion and metastasis of lung cancer - Gu et al., 2014)
8. TargetScan microRNA MIR495 (mimics of this miRNA inhibit lung tumor progression - Ai et al., 2015)
9. ENCODE TF ChIP-seq 2015 - HA-E2F1_MCF-7_hg19, HA-E2F1_HeLa-S3_hg19, E2F1_HeLA-Se_hg19, MYC, YY1, CEBPB  
10. TF-LOF Expression from GEO - gata3, yy1, myb, znf217, gata3

### Notes on enrichr results for genes down with COPD+cancer vs COPD  
1. [NOTE - pathways not working at this time]  
2. ChEA - EGR1, SOX2, SPI1, MITF, GATA2, TP63, FOXP1, MYC, YY1, SPI1  
3. TRANSFAC/JASPAR - WT1, ZNF148, TFAP2A, SP3, KLF4, KLF5, SP1, NFKB1  
4. ENCODE TF ChIP-seq 2015 - SPI1, 


### Test ZNF217 expression between cancers  
```{r ZNF217a}
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "FinalCaDXc", "0")

ind <- match("7764_at", featureNames(eset))
summary(lm(exprs(eset)[ind, ] ~ eset$COPD2_R7 + eset$SMKc + eset$AllegroCorticosteroidsc + eset$COPD2_R7:eset$AllegroCorticosteroidsc))

```

### Test ZNF217 expression between COPD and COPD+Cancer
```{r ZNF217b}
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "0")

ind <- match("7764_at", featureNames(eset))
summary(lm(exprs(eset)[ind, ] ~ eset$FinalCaDXc + eset$SMKc + eset$AllegroCorticosteroidsc))


```

### Test EGR1 expression between cancers
```{r EGR1a}
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "FinalCaDXc", "0")

ind <- match("1958_at", featureNames(eset))
summary(lm(exprs(eset)[ind, ] ~ eset$COPD2_R7 + eset$SMKc + eset$AllegroCorticosteroidsc + eset$COPD2_R7:eset$AllegroCorticosteroidsc))


```

### Test EGR1 expression between COPD and COPD+Cancer
```{r EGR1b}
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "0")

ind <- match("1958_at", featureNames(eset))
summary(lm(exprs(eset)[ind, ] ~ eset$FinalCaDXc + eset$SMKc + eset$AllegroCorticosteroidsc + eset$FinalCaDXc:eset$AllegroCorticosteroidsc))


```

### Test FN1 expression between the groups
```{r FN1}
# test between COPD and COPD+Cancer
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "0")

ind <- match("2335_at", featureNames(eset))
summary(lm(exprs(eset)[ind, ] ~ eset$FinalCaDXc + eset$AGEcalc + eset$RIN + eset$SMKc + eset$AllegroCorticosteroidsc + eset$FinalCaDXc:eset$AllegroCorticosteroidsc))

# test between Control and COPD+Cancer
eset <- removeFactorLevel(removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "indicator", "2"), "indicator", "3")

summary(lm(exprs(eset)[ind, ] ~ eset$indicator + eset$AGEcalc + eset$RIN + eset$SMKc + eset$AllegroCorticosteroidsc))


# make sure of directionality
t.test(exprs(eset)[ind, eset$indicator==1], exprs(eset)[ind, eset$indicator==4])

# test between Control and COPD
eset <- removeFactorLevel(removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "indicator", "2"), "indicator", "4")

summary(lm(exprs(eset)[ind, ] ~ eset$indicator + eset$SMKc + eset$AllegroCorticosteroidsc))

# test between Control and Cancer
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "1")

summary(lm(exprs(eset)[ind, ] ~ eset$FinalCaDXc + eset$SMKc + eset$AllegroCorticosteroidsc + eset$FinalCaDXc:eset$AllegroCorticosteroidsc))


```

### Test LAMP3 Expression - expect increased in COPD and COPD+Cancer (resulting from ATF4 role)
```{r LAMP3}
# test between COPD and COPD+Cancer
eset <- removeFactorLevel(removeFactorLevel(removeBioReps(esetCleanFormers), "AllegroCorticosteroidsc", "DK"), "FinalCaDXc", "1")

ind <- match("27074_at", featureNames(eset))
summary(lm(exprs(eset)[ind, ] ~ eset$COPD2_R7 + eset$AllegroCorticosteroidsc))


```

### What is changing with COPD severity in COPD and in COPD+Cancer?
```{r severityAndCancer}

# try using both the PFT measurements and also a projected COPD severity genomic score using Steiling98 and/or Campbell127  
eset <- removeFactorLevel(removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "0"), "FinalCaDXc", "1")

#beeswarm(eset$RATIOc ~ eset$FinalCaDXc)

severityCancer <- lmFitWrapper(eset, c("RATIOc", "GENDERc", "SMKc", "RIN"), 1, "none", 0.05)
o = order(eset$FEV1c, decreasing=TRUE)
eset = eset[, o]
heatmap3(exprs(eset)[severityCancer$inds, ], col=bluered,
         hclustfun=function(d) hclust(d, method="ward.D"),
         labCol="", ColSideColors=gender_colors[eset$GENDERc])

eset <- removeFactorLevel(removeFactorLevel(removeFactorLevel(removeBioReps(esetClean), "AllegroCorticosteroidsc", "DK"), "COPD2_R7", "0"), "FinalCaDXc", "0")

severityCancer2 <- lmFitWrapper(eset, c("RATIOc", "GENDERc", "SMKc", "RIN", "AllegroCorticosteroidsc"), 1, "none", 0.05)

oppGenes <- severityCancer$fit$genes$Symbol[which(severityCancer2$fit$t[,2] < -1 &severityCancer$fit$t[,2] > 1)]

write.table(oppGenes, file="oppGenes20150717.txt", quote=FALSE, row.names=FALSE, col.names=FALSE)

```